text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+However, this RFC doesn't actually stop `uN` and `iN` from being added as aliases to `u<N>` and `i<N>` in the future. While the version with an explicit generic parameter is required for generic code, it should in theory be possible to add these aliases if people want them. These would require special changes to the language itself, since it does not make sense to add `2.pow(32)` items to the prelude.
+
+## Use `uint<N>` and `int<N>` instead of `u<N>` and `i<N>`
+
+This was the previous proposal used by the original RFC, and they were chosen due to a misconception about name conflicts. Although the `i` variable is commonly used for loops and other index variables, variable names and type names are entirely separate and can coexist peacefully. Having names closer to the originals feels more discoverable here.
+
+Additionally, the names `uint` and `int` remain available for future, potentially unrelated integer types.
+
+## Bound-based generalisation
+
+Generalising integers over their number of bits is certainly a very logical way to generalise integers *for computers*, but generalising based upon bounds is a very natural way for humans to do it, *and* more general. For example, instead of `u<N>` and `i<N>` types, we could get away with just one type,
+`int<A..=B>`. This would be more powerful than the original: for example, a percentage could be represented exactly as `int<0..=100>`. Whether an integer is signed simply depends on whether its lower bound is negative.
+
+The primary reason for leaving this out is… well, it's a lot different from the existing integer types in the language. Additionally, such a proposal can coexist with this one, since there's nothing stopping us from making an `int<A..=B>` type in the future such that `u<N>` and `i<N>` are equivalent to the correct range of integers.
+
+## Integer traits
+
+Previously, Rust had traits which generalized the integer types and their methods, but these were ultimately removed in favour of inherent methods. Going with a generic `u<N>` over an `Int` trait would avoid the problem of determining which methods are suited for these traits; instead, the `u<N>` type would have all of them.
+
+Additionally, having these traits does not allow non-power-of-two `u<N>` outright, and so, this feature is a strict superset of that.
+
+Having separate `u<N>` and `i<N>` types also solves the problem of generalising signed and unsigned integers: everything would require one impl for signed integers, and one for unsigned. This would ensure that these types have exactly the same behaviour for all `N`, only differing in the upper bound on the number of bits.
+
+## Offering as a library
+
+This was the main proposal last time this RFC rolled around, and as we've seen, it hasn't really worked.
+
+Crates like [`u`], [`bounded-integer`], and [`intx`] exist, but they come with their own host of problems:
+
+* None of these libraries can easily unify with the existing `uN` and `iN` types.
```

Comment:
A const-generic library type can't provide this and also can't support literals. But what problems exactly does that cause in practice? Which aspects can be handled well with existing language features and which ones really need language support?

Reply:
The RFC already mentions how being able to provide a small number of generic impls that cover all integer types has an extremely large benefit over being forced to use macros to implement for all of them individually. You cannot do this without language support.

Reply:
So this bullet point is "only" about impls like `impl<const BITS: usize> Foo for some_library::Int<BITS> { ... }` not implementing anything for the primitive integer types? Could `From` impls and some form of delegation (#3530) also help with this?

Reply:
Not really, and this is mentioned in the RFC also. That's 5 impls for unsigned, 5 impls for signed that could just be 2 impls, whether you have delegation or not. Even for simple traits, like `Default`, you're incentivised to use a macro just because it becomes so cumbersome.

Reply:
`arbitrary-int` provides a unification somewhat using its `Number` trait. It's somewhat rudimentary but I am working on improving it.

Reply:
Reading this again, the Number trait fulfills a somewhat different role though. It allows writing generic code against any Number (be it an arbitrary-int or a native int), but it does not expose the bits itself - which can be a plus or a minus, depending on what you're building.